seed/seedwriter/seed20.go

```diff
@@ -0,0 +1,322 @@
+}
+
+func (pol *policy20) needsImplicitSnaps(*naming.SnapSet) (bool, error) {
+	// no implicit snaps with Core 20
+	// TODO: unless we want to support them for extra snaps
+	return false, nil
+}
+
+func (pol *policy20) implicitSnaps(*naming.SnapSet) []*asserts.ModelSnap {
+	return nil
+}
+
+func (pol *policy20) implicitExtraSnaps(*naming.SnapSet) []*OptionsSnap {
+	return nil
+}
+
+type tree20 struct {
+	opts *Options
+
+	snapsDirPath string
+	systemDir    string
+
+	systemSnapsDirEnsured bool
+}
+
+func (tr *tree20) mkFixedDirs() error {
+	tr.snapsDirPath = filepath.Join(tr.opts.SeedDir, "snaps")
+	tr.systemDir = filepath.Join(tr.opts.SeedDir, "systems", tr.opts.Label)
+
+	if err := os.MkdirAll(tr.snapsDirPath, 0755); err != nil {
+		return err
+	}
+
+	return os.MkdirAll(tr.systemDir, 0755)
+}
+
+func (tr *tree20) ensureSystemSnapsDir() error {
+	if tr.systemSnapsDirEnsured {
+		return nil
+	}
+	snapsDir := filepath.Join(tr.systemDir, "snaps")
+	if err := os.MkdirAll(snapsDir, 0755); err != nil {
+		return err
+	}
+	tr.systemSnapsDirEnsured = true
+	return nil
+}
+
+func (tr *tree20) snapsDir() string {
+	// XXX what about extra snaps?
+	return tr.snapsDirPath
+}
+
+func (tr *tree20) localSnapPath(*SeedSnap) string {
+	panic("XXX dangerous features not implemented yet")
+}
+
+func (tr *tree20) writeAssertions(db asserts.RODatabase, modelRefs []*asserts.Ref, snapsFromModel []*SeedSnap, extraSnaps []*SeedSnap) error {
+	assertsDir := filepath.Join(tr.systemDir, "assertions")
+	if err := os.MkdirAll(assertsDir, 0755); err != nil {
+		return err
+	}
+
+	writeByRefs := func(fname string, refsGen func(stop <-chan struct{}) <-chan *asserts.Ref) error {
+		f, err := os.OpenFile(filepath.Join(assertsDir, fname), os.O_WRONLY|os.O_CREATE|os.O_EXCL, 0644)
+		if err != nil {
+			return err
+		}
+		defer f.Close()
+
+		stop := make(chan struct{})
+		defer close(stop)
+		refs := refsGen(stop)
+
+		enc := asserts.NewEncoder(f)
+		for {
+			aRef := <-refs
+			if aRef == nil {
+				break
+			}
+			a, err := aRef.Resolve(db.Find)
+			if err != nil {
+				return fmt.Errorf("internal error: lost saved assertion")
+			}
+			if err := enc.Encode(a); err != nil {
+				return err
+			}
+		}
+		return nil
+	}
+
+	pushRef := func(refs chan<- *asserts.Ref, ref *asserts.Ref, stop <-chan struct{}) bool {
+		select {
+		case refs <- ref:
+			return true
+		case <-stop:
+			// get unstuck if we error early
+			return false
+		}
+	}
+
+	modelOnly := func(aRef *asserts.Ref) bool { return aRef.Type == asserts.ModelType }
+	excludeModel := func(aRef *asserts.Ref) bool { return aRef.Type != asserts.ModelType }
+
+	modelRefsGen := func(include func(*asserts.Ref) bool) func(stop <-chan struct{}) <-chan *asserts.Ref {
+		return func(stop <-chan struct{}) <-chan *asserts.Ref {
+			refs := make(chan *asserts.Ref)
+			go func() {
+				for _, aRef := range modelRefs {
+					if !include(aRef) {
+						continue
+					}
+					if !pushRef(refs, aRef, stop) {
+						return
+					}
```

Comment:
this is fine for now so you can land while green, but something to consider for a followup is just:

```go
					if include(aRef) {
						if !pushRef(refs, aRef, stop) {
							return
						}	
					}
```

the continue here is unnecessary if we put the pushRef inside the if condition